File: 0001-0500/0042/0042.trapping-rain-water.cpp
/*
 * @lc app=leetcode id=42 lang=cpp
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
class Solution
{
public:
    int trap(vector<int> &height)
    {
        int n = height.size();
        if (n == 0)
            return 0;

        int left = 0, right = n - 1;
        int leftMax = height[left], rightMax = height[right];
        int res = 0;

        while (left < right)
        {
            if (height[left] < height[right])
            {
                if (height[left] < leftMax)
                {
                    res += leftMax - height[left];
                }
                else
                {
                    leftMax = height[left];
                }
                left++;
            }
            else
            {
                if (height[right] < rightMax)
                {
                    res += rightMax - height[right];
                }
                else
                {
                    rightMax = height[right];
                }
                right--;
            }
        }
        return res;
    }
};
// @lc code=end

File: 0001-0500/0042/0042.trapping-rain-water.js
/*
 * @lc app=leetcode id=42 lang=javascript
 *
 * [42] Trapping Rain Water
 */

// @lc code=start
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
    let n = height.length;
    let left = new Array(n);
    let right = new Array(n);
    let ans = 0;
    left[0] = height[0];
    for (let i = 1; i < n; i++) {
        left[i] = Math.max(left[i - 1], height[i]);
    }
    right[n - 1] = height[n - 1];
    for (let i = n - 2; i >= 0; i--) {
        right[i] = Math.max(right[i + 1], height[i]);
    }
    for (let i = 0; i < n; i++) {
        ans += Math.min(left[i], right[i]) - height[i];
    }
    return ans;
};
// @lc code=end

